### ISSUE TITLE
Snowflake Sink committing offsets for null/tombstone messages

### ISSUE THREAD

**Issue by lsimac** [snowflakedb/snowflake-kafka-connector#756]
Hi everyone,

We've observed that when `behavior.on.null` is set to `IGNORE` (meaning null messages are not inserted into Snowflake), offsets are not committed for those null messages. I believe it's worth considering committing offsets for null messages, even in cases where they are not inserted into Snowflake. This approach maintains consistency, as null messages are actually consumed by the Snowflake sink connector (though they are not sent to Snowflake in the end).

This adjustment would address the issue where null messages get 'stuck' in the topic, triggering alerts about Consumer (Kafka Connect) lagging or being too slow. During this time, the lag doesn't decrease because Kafka Connect is continuously pulling those null Kafka Messages without committing offsets for them. The lag/alert is resolved only when a message with a non-null value enters the Kafka topic, prompting the Snowflake sink to commit the offset for it.

**Comment by sfc-gh-xhuang** [comment]
@sfc-gh-rcheng can you chime in here?

**Comment by sfc-gh-rcheng** [comment]
Hi @lsimac this sounds like you're looking for tombstone messages, which are empty rows committed into snowflake with just the offset and key. Tombstone ingestion can be enabled by setting `behavior.on.null` to `DEFAULT` after v1.5.5 for Snowpipe and v2.1.0 for Snowpipe Streaming.

Would this work for you? 

In regards to the null messages causing consumer lag in Kafka, the offsets fetched by KC are pulled from the latest committed offset in Snowflake. Unfortunately this means that the offset must be committed with data or as a tombstone, we cannot advance or commit an offset without actual data being sent.

**Comment by lsimac** [comment]
Hi @sfc-gh-rcheng,

Thank you for your response. In our scenario, setting `behavior.on.null` to DEFAULT is not viable as we aim to avoid the insertion of tombstone messages. My initial comment leaned towards the possibility of modifying the ingestion logic for handling null messages (feature request). The reason behind this suggestion, as you pointed out, is that the current offsets from Kafka Connect primarily rely on Snowflake offsets. Given that null messages are not stored in Snowflake when `behavior.on.null` is set to IGNORE, consumer lag is reported for null messages when a null message is the last in the Kafka partition.

I propose a modification in the logic, wherein offsets could be pushed or set to Snowflake (if SDK supports that) for null messages, even if these messages won't be stored in Snowflake. This change could be made for instance when a null message is the last in the batch. I believe this is a valid approach because the Snowflake sink successfully receives null messages therefore it makes sense to record the offset of a null message in Snowflake (or at least report the offset back to Kafka) even if the null message itself is not going to be inserted into Snowflake.

To summarize, while inserting null messages into Snowflake is not an option for us, we would still like the offsets of null messages to be committed back to Kafka so that we do not have consumer lag.

**Comment by sfc-gh-rcheng** [comment]
@lsimac I see thank you for the explanation. Your goal is to bump the offset in the Snowflake sink when a null message is sent so that the Kafka consumer lag is not increased. Unfortunately this is not possible with our current architecture

The Snowflake sink Kafka Connector PUTs and PRECOMMITs data. 

The PUT is what you referenced in 'Snowflake sink successfully receives null messages' that receives and processes data using the [processedOffset](https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java#L100). This offset is bumped when a null message is received. 

The PRECOMMIT on the other hand is an entirely different code path that gets the [offsetPersistedInSnowflake](https://github.com/snowflakedb/snowflake-kafka-connector/blob/master/src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java#L94) directly from the last record sent to Snowflake servers, separate from the Snowflake sink. This cannot be bumped by the Snowflake sink because it is pulled directly from the last non-null or tombstone record sent to Snowflake, it is not a counter that we keep server side.

We need to keep these two offsets on different code paths to ensure exactly once delivery. This means that all messages committed back to Kafka with the PRECOMMIT code path must be ingested into Snowflake, we cannot bump the offset server side without a record to reduce consumer lag.

Let me know if you have further questions

**Comment by sfc-gh-akowalczyk** [comment]
Closing due to inactivity in this topic.

**Comment by simonepm** [comment]
Hi @sfc-gh-rcheng I see "Unfortunately this is not possible with our current architecture", do you think is possible to have it as a connector evolution feature? e.g adding a flag like IGNORE_AND_COMMIT to the existing DEFAULT, IGNORE and ERROR.

I say this as having data anyway landing as null on Snowflake side produce higher costs with no added benefit, and for some security requirements certain records should not land at all, even if nullified.

"This cannot be bumped by the Snowflake sink because it is pulled directly from the last non-null or tombstone record sent to Snowflake" I see the point, do you think adding a fetch from offset stored in Kafka is something possible?

Thank you!
